### PR TITLE
Avoid repetitive GCHandles in IIS WebSocket operations

### PR DESCRIPTION
**Pull request by BrennanConroy**
Make the `AsyncIOEngine` in IIS disposable so we can avoid creating a new `GCHandle` for every WebSocket read and write operation.

### CODE REVIEW

File: src/Servers/IIS/IIS/src/Core/IO/WebSocketsAsyncIOEngine.Write.cs [hunk 1]
```diff
@@ -24,26 +24,30 @@ private static NativeMethods.REQUEST_NOTIFICATION_STATUS WriteCallback(IntPtr ht
...
 
         protected override unsafe int WriteChunks(NativeSafeHandle requestHandler, int chunkCount, HttpApiTypes.HTTP_DATA_CHUNK* dataChunks, out bool completionExpected)
         {
-            _thisHandle = GCHandle.Alloc(this);
             return NativeMethods.HttpWebsocketsWriteBytes(requestHandler, dataChunks, chunkCount, &WriteCallback, (IntPtr)_thisHandle, out completionExpected);
         }
 
         protected override void ResetOperation()
         {
             base.ResetOperation();
```
Comment (BrennanConroy) [inline src/Servers/IIS/IIS/src/Core/IO/WebSocketsAsyncIOEngine.Write.cs:43]:
```suggestion
```

**Comment by Tratcher** [comment]
What's the impact? Is it just allocations, or is there some perf cost to making the GC handle over and over?

**Review by Tratcher — APPROVED** [review]

**Comment by davidfowl** [comment]
There’s a performance cost yes

**Comment by davidfowl** [comment]
Do the uncached operations get disposed?

**Comment by BrennanConroy** [comment]
> Do the uncached operations get disposed?

It shouldn't be possible to get uncached ones right?
The read and write are in single loops
https://source.dot.net/#Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.IO.cs,105

**Comment by davidfowl** [comment]
> It shouldn't be possible to get uncached ones right?
The read and write are in single loops
https://source.dot.net/#Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.IO.cs,105

That's technically true, but the code isn't written that way.

**Comment by BrennanConroy** [comment]
Would removing the `Get*Operation()` and `ReturnOperation()` methods make that clearer?

**Comment by davidfowl** [comment]
I think so?

**Comment by BrennanConroy** [comment]
Thoughts on new changes?

**Comment by davidfowl** [comment]
This looks much easier to follow!

**Review by Tratcher — APPROVED** [review]

**Pull request merged**